File: src/main/java/net/atos/zac/admin/ZaaktypeCmmnConfigurationService.java
/*
 * SPDX-FileCopyrightText: 2021 Atos, 2024 INFO.nl
 * SPDX-License-Identifier: EUPL-1.2+
 */

package net.atos.zac.admin;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.TimeUnit;
import java.util.logging.Logger;
import java.util.stream.Collectors;

import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;

import com.github.benmanes.caffeine.cache.Cache;
import com.github.benmanes.caffeine.cache.Caffeine;
import com.github.benmanes.caffeine.cache.RemovalCause;
import com.github.benmanes.caffeine.cache.stats.CacheStats;

import net.atos.client.zgw.shared.cache.Caching;
import net.atos.zac.admin.model.ZaaktypeCmmnConfiguration;
import nl.info.zac.admin.ZaaktypeCmmnConfigurationBeheerService;

@ApplicationScoped
public class ZaaktypeCmmnConfigurationService implements Caching {
    private static final Logger LOG = Logger.getLogger(ZaaktypeCmmnConfigurationService.class.getName());
    private static final int MAX_CACHE_SIZE = 20;
    private static final int EXPIRATION_TIME_HOURS = 1;

    /**
     * Hardcoded zaakbeeindigreden that we don't manage via ZaaktypeCmmnConfiguration
     */
    public static final String INADMISSIBLE_TERMINATION_ID = "ZAAK_NIET_ONTVANKELIJK";
    public static final String INADMISSIBLE_TERMINATION_REASON = "Zaak is niet ontvankelijk";

    @Inject
    private ZaaktypeCmmnConfigurationBeheerService zaaktypeCmmnConfigurationBeheerService;

    private static final Map<String, Cache<?, ?>> CACHES = new HashMap<>();

    private <K, V> Cache<K, V> createCache(String name) {
        Cache<K, V> cache = Caffeine.newBuilder()
                .maximumSize(MAX_CACHE_SIZE)
                .expireAfterAccess(EXPIRATION_TIME_HOURS, TimeUnit.HOURS)
                .recordStats()
                .removalListener(
                        (K key, V value, RemovalCause cause) -> LOG.fine("Removing key: %s in cache %s because of: %s".formatted(key, name,
                                cause))
                )
                .build();

        CACHES.put(name, cache);
        return cache;
    }

    private final Cache<UUID, ZaaktypeCmmnConfiguration> uuidToZaaktypeCmmnConfigurationCache = createCache(
            "UUID -> ZaaktypeCmmnConfiguration");
    private final Cache<String, List<ZaaktypeCmmnConfiguration>> stringToZaaktypeCmmnConfigurationListCache = createCache(
            "List<ZaaktypeCmmnConfiguration>");

    public ZaaktypeCmmnConfiguration readZaaktypeCmmnConfiguration(final UUID zaaktypeUUID) {
        return uuidToZaaktypeCmmnConfigurationCache.get(
                zaaktypeUUID,
                uuid -> zaaktypeCmmnConfigurationBeheerService.fetchZaaktypeCmmnConfiguration(zaaktypeUUID)
        );
    }

    public List<ZaaktypeCmmnConfiguration> listZaaktypeCmmnConfiguration() {
        return stringToZaaktypeCmmnConfigurationListCache.get(
                ZAC_ZAAKTYPECMMNCONFIGURATION,
                s -> zaaktypeCmmnConfigurationBeheerService.listZaaktypeCmmnConfiguration()
        );
    }

    public boolean isSmartDocumentsEnabled(final UUID zaaktypeUUID) {
        return readZaaktypeCmmnConfiguration(zaaktypeUUID).isSmartDocumentsIngeschakeld();
    }

    public void cacheRemoveZaaktypeCmmnConfiguration(final UUID zaaktypeUUID) {
        uuidToZaaktypeCmmnConfigurationCache.invalidate(zaaktypeUUID);
    }

    public String clearManagedCache() {
        uuidToZaaktypeCmmnConfigurationCache.invalidateAll();
        return cleared(Caching.ZAC_ZAAKTYPECMMNCONFIGURATION_MANAGED);
    }

    public String clearListCache() {
        stringToZaaktypeCmmnConfigurationListCache.invalidateAll();
        return cleared(Caching.ZAC_ZAAKTYPECMMNCONFIGURATION);
    }

    @Override
    public Map<String, CacheStats> cacheStatistics() {
        return CACHES.entrySet()
                .stream()
                .collect(Collectors.toMap(Map.Entry::getKey, cache -> cache.getValue().stats()));
    }

    @Override
    public Map<String, Long> estimatedCacheSizes() {
        return CACHES.entrySet()
                .stream()
                .collect(Collectors.toMap(Map.Entry::getKey, cache -> cache.getValue().estimatedSize()));
    }
}

File: src/main/kotlin/nl/info/zac/admin/ZaaktypeCmmnConfigurationService.kt
/*
 * SPDX-FileCopyrightText: 2021 Atos, 2026 INFO.nl
 * SPDX-License-Identifier: EUPL-1.2+
 */
package nl.info.zac.admin

import com.github.benmanes.caffeine.cache.Cache
import com.github.benmanes.caffeine.cache.Caffeine
import com.github.benmanes.caffeine.cache.stats.CacheStats
import jakarta.enterprise.context.ApplicationScoped
import jakarta.inject.Inject
import net.atos.client.zgw.shared.cache.Caching
import nl.info.zac.admin.model.ZaaktypeCmmnConfiguration
import nl.info.zac.util.AllOpen
import nl.info.zac.util.NoArgConstructor
import java.util.UUID
import java.util.concurrent.TimeUnit
import java.util.logging.Logger

@ApplicationScoped
@AllOpen
@NoArgConstructor
class ZaaktypeCmmnConfigurationService @Inject constructor(
    private val zaaktypeCmmnConfigurationBeheerService: ZaaktypeCmmnConfigurationBeheerService
) : Caching {

    companion object {
        const val INADMISSIBLE_TERMINATION_ID = "ZAAK_NIET_ONTVANKELIJK"
        const val INADMISSIBLE_TERMINATION_REASON = "Zaak is niet ontvankelijk"

        private val LOG = Logger.getLogger(ZaaktypeCmmnConfigurationService::class.java.name)
        private const val MAX_CACHE_SIZE = 20L
        private const val EXPIRATION_TIME_HOURS = 1L

        private val caches = mutableMapOf<String, Cache<*, *>>()

        private fun <K : Any, V : Any> createCache(name: String): Cache<K, V> =
            Caffeine.newBuilder()
                .maximumSize(MAX_CACHE_SIZE)
                .expireAfterAccess(EXPIRATION_TIME_HOURS, TimeUnit.HOURS)
                .recordStats()
                .removalListener<K, V> { key, _, cause ->
                    LOG.fine("Removing key: $key in cache $name because of: $cause")
                }
                .build<K, V>()
                .also { caches[name] = it }
    }

    private val uuidToConfigCache: Cache<UUID, ZaaktypeCmmnConfiguration> =
        createCache("UUID -> ZaaktypeCmmnConfiguration")

    private val listCache: Cache<String, List<ZaaktypeCmmnConfiguration>> =
        createCache("List<ZaaktypeCmmnConfiguration>")

    fun readZaaktypeCmmnConfiguration(zaaktypeUUID: UUID): ZaaktypeCmmnConfiguration =
        uuidToConfigCache.get(zaaktypeUUID) {
            zaaktypeCmmnConfigurationBeheerService.fetchZaaktypeCmmnConfiguration(it)
        }

    fun listZaaktypeCmmnConfiguration(): List<ZaaktypeCmmnConfiguration> =
        listCache.get(Caching.ZAC_ZAAKTYPECMMNCONFIGURATION) {
            zaaktypeCmmnConfigurationBeheerService.listZaaktypeCmmnConfiguration()
        }

    fun cacheRemoveZaaktypeCmmnConfiguration(zaaktypeUUID: UUID) {
        uuidToConfigCache.invalidate(zaaktypeUUID)
    }

    fun clearManagedCache(): String {
        uuidToConfigCache.invalidateAll()
        return cleared(Caching.ZAC_ZAAKTYPECMMNCONFIGURATION_MANAGED)
    }

    fun clearListCache(): String {
        listCache.invalidateAll()
        return cleared(Caching.ZAC_ZAAKTYPECMMNCONFIGURATION)
    }

    override fun cacheStatistics(): Map<String, CacheStats> =
        caches.mapValues { it.value.stats() }

    override fun estimatedCacheSizes(): Map<String, Long> =
        caches.mapValues { it.value.estimatedSize() }
}
